assets/js/pixels.js

```diff
@@ -3,42 +3,207 @@
         }
     }
+
+    // Helper: requestTick
+    // Better performance on scrolling by using debouncing.
+    // Reference: https://gomakethings.com/debouncing-events-with-requestanimationframe-for-better-performance/
+    function requestTick() {
+        if (!ticking) {
+            requestAnimationFrame(updateScrollInfo);
+        }
+        ticking = true;
+    }
+
+    // Helper: updateScrollInfo
+    // ---------------
+    // Check which images are now in view.
+    function updateScrollInfo() {
+        let progressMax = lastDocumentHeight - lastWindowHeight;
+        let percentage = lastScrollY / progressMax;
+
+        $articleImagesEl.each(function(index) {
+            if (elementIsVisibleInViewport(this, true)) {
+                firePixel('image', source, pathname,
+                          (index + 1) + 'of' + $articleImagesEl.length,
+                          {once: true});
+            }
+        });
+
+        ticking = false;
+    }
+
+    // Helper: sanitizeUrl
+    // Sanitzes a given URL by removing unsafe characters.
+    // E.g. 'https://reddit.com/r/duckduckgo' -> 'https-reddit-com-r-duckduckgo'
+    function sanitizeUrl(url) {
+        return url
+        // strip leading/trailing slash
+            .replace(/^\/|\/$/, '')
+        // strip unsafe chars for grafana
+            .replace(/[^a-z0-9_-]+/ig, '-')
+        // strip underscores as well
+            .replace(/_/g,'-');
+    }
+
+    // Helper: elementIsVisibleInViewport
+    // Source: https://github.com/30-seconds/30-seconds-of-code#elementisvisibleinviewport-
+    // Figures out if an element (in our case images) are in view so that we
+    // can fire the pixels.
+    function elementIsVisibleInViewport(el, partiallyVisible) {
+        partiallyVisible = partiallyVisible || false;
+
+        const elBounds = el.getBoundingClientRect();
+        const top = elBounds.top;
+        const left = elBounds.left;
+        const bottom = elBounds.bottom;
+        const right = elBounds.right;
+
+        const innerHeight = window.innerHeight;
+        const innerWidth = window.innerWidth;
+
+        if (partiallyVisible) {
+            return ((top > 0 && top < innerHeight) ||
+                    (bottom > 0 && bottom < innerHeight)) &&
+                   ((left > 0 && left < innerWidth) ||
+                    (right > 0 && right < innerWidth));
+        } else {
+            return top >= 0 &&
+                   left >= 0 &&
+                   bottom <= innerHeight &&
+                   right <= innerWidth;
+        }
+    }
+
+    // --------------------
+    // III. Event Listeners
+    // --------------------
+
+    // Page Loading
+    // ------------
+    // Fire when the post loads.
+    // Example: blog.load.quora.google-filter-bubble-study
+    firePixel(pixels.position.load, source, pathname, {once: true});
+
+    // Page Position Scrolling
+    // -----------------------
+    // Figure out if the user is 25%, 50%, or 100% of the page.
```

Comment:
Let's update this to explain we're trying to detect when images are viewed

Reply:
👍🏽